sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -824,59 +944,158 @@ case class HashAggregateExec(
           val effectiveCodes = subExprs.codes.mkString("\n")
-          val fastRowEvals = ctx.withSubExprEliminationExprs(subExprs.states) {
-            boundUpdateExpr.map(_.genCode(ctx))
+          val fastRowEvals = boundUpdateExprs.map { boundUpdateExprsForOneFunc =>
+            ctx.withSubExprEliminationExprs(subExprs.states) {
+              boundUpdateExprsForOneFunc.map(_.genCode(ctx))
+            }
           }
-          val updateFastRow = fastRowEvals.zipWithIndex.map { case (ev, i) =>
-            val dt = updateExpr(i).dataType
-            CodeGenerator.updateColumn(
-              fastRowBuffer, dt, i, ev, updateExpr(i).nullable, isVectorized = true)
+
+          val aggCodeBlocks = fastRowEvals.zipWithIndex.map { case (fastRowEvalsForOneFunc, i) =>
+            val boundUpdateExprsForOneFunc = boundUpdateExprs(i)
+            val bufferOffset = bufferStartOffsets(i)
+            // All the update code for aggregation buffers should be placed in the end
+            // of each aggregation function code.
+             val updateRowBuffer = fastRowEvalsForOneFunc.zipWithIndex.map { case (ev, j) =>
+               val updateExpr = boundUpdateExprsForOneFunc(j)
+               val dt = updateExpr.dataType
+               val nullable = updateExpr.nullable
+               CodeGenerator.updateColumn(fastRowBuffer, dt, bufferOffset + j, ev, nullable,
+                 isVectorized = true)
+             }
+             code"""
+                |// evaluate aggregate function for ${aggNames(i)}
+                |${evaluateVariables(fastRowEvalsForOneFunc)}
+                |// update fast row
+                |${updateRowBuffer.mkString("\n").trim}
+              """.stripMargin
           }
 
-          // If vectorized fast hash map is on, we first generate code to update row
-          // in vectorized fast hash map, if the previous loop up hit vectorized fast hash map.
-          // Otherwise, update row in regular hash map.
-          s"""
-             |if ($fastRowBuffer != null) {
-             |  // common sub-expressions
-             |  $effectiveCodes
-             |  // evaluate aggregate function
-             |  ${evaluateVariables(fastRowEvals)}
-             |  // update fast row
-             |  ${updateFastRow.mkString("\n").trim}
-             |} else {
-             |  $updateRowInRegularHashMap
-             |}
-          """.stripMargin
+          lazy val nonSplitAggCode = {
+            // If vectorized fast hash map is on, we first generate code to update row
+            // in vectorized fast hash map, if the previous loop up hit vectorized fast hash map.
+            // Otherwise, update row in regular hash map.
+            s"""
+               |if ($fastRowBuffer != null) {
+               |  // common sub-expressions
+               |  $effectiveCodes
+               |  // evaluate aggregate functions and update aggregation buffers
+               |  ${aggCodeBlocks.fold(EmptyBlock)(_ + _)}
```

Comment:
only this line is different if we split. How about
```
val codeToEvalAggFunc = if (conf.codegenSplitAggregateFunc && ... ) {
  ...
  maybeSplitCode.getOrElse(aggCodeBlocks.fold(EmptyBlock)(_ + _))
} else {
  aggCodeBlocks.fold(EmptyBlock)(_ + _)
}
s"""
  |...
  |$codeToEvalAggFunc
  """

```

Reply:
nice suggestion and I'll follow it.